### DISCUSSION TITLE
Link to other pages

### DISCUSSION THREAD

**Discussion by Bernardstanislas** [11ty/eleventy#1795] [Q&A] [answered]
### Problem statement

Sorry for the noob question but I can't figure out how to correctly model my use case.

I am creating a food blog, with some recipes. Each recipe has the same structure, let's say a `title`, a `picture` and an array of `steps`.
Each `step` has a `picture`, and a `content`.

Next to my recipes, I also have one page per technical task. A technical task is rendered on its own.

**My problem:** in the `content` of a `step` included in a `recipe`, I would like to create put link to a technical task page, when needed. I cannot figure out how to dynamically transform a text inside the `content` property into a dynamically crafter link to another page.

### Example

Let's say I have a macarons recipe:

```js
// _data/recipes.js
module.exports = [
  {
    title: 'Macarons',
    picture: 'macarons.jpg',
    steps: [
      {
        picture: 'step1.jpg',
        content: 'Take the eggs and [scramble]({{ technical_tasks.egg_scrambling}}) them.'
      },
// [...]
```

In the `content` of my first step, I do not know how I can create a link.

Any thoughts on that?

Thanks a lot!

**Comment by pdehaan** [comment, marked as answer]
@Bernardstanislas Would a hardcoded link work?

```js
content: 'Take the eggs and <a href="/technical/scramble/">scramble</a> them.'
```

To be honest I'm not sure if the markdown+liquid/nunjucks template syntax would work within a global data file. 

I'm not sure what `technical_tasks` is, or where that comes from, but if it's another global data file of links, you might need to include that file in your `./_data/recipes.js` file.

**Reply by Bernardstanislas** [reply]
Yes a hardcoded link would definitely work, I was looking for a way to get the technical task permalink.

I'll go for this solution then, thanks a lot for your help 🙇

**Reply by pdehaan** [reply]
OK, this seems to work locally for me&hellip;

```js
// _data/recipes.js
// NOTE: Added the require to ./_data/technical_files.js here so I can use string interpolation below.
const technical_tasks = require("./technical_tasks");

module.exports = [
  {
    title: 'Macarons',
    picture: 'macarons.jpg',
    steps: [
      {
        picture: 'step1.jpg',
        // NOTE: The backticks and JavaScript template string syntax here.
        content: `Take the eggs and [scramble](${ technical_tasks.egg_scrambling }) them.`
      },
    ]
  }
];
```

```js
// _data/technical_tasks.js
module.exports = {
  egg_scrambling: "/technical/scramble-eggs/",
};
```

```md
<!-- ./index.md -->
{% for recipe in recipes %}
  # {{ recipe.title }}
  {% for step in recipe.steps %}
  1. {{ step.content }}
  {% endfor %}
{% endfor %}
```

And my output is:

```html
<h1>Macarons</h1>
<ol>
  <li>Take the eggs and <a href="/technical/scramble-eggs/">scramble</a> them.</li>
</ol>
```

Depending on how deep your HTML-in-markdown nesting, I found this pitfall in the docs very useful: https://www.11ty.dev/docs/languages/markdown/#there-are-extra-and-in-my-output

**Reply by Bernardstanislas** [reply]
Brilliant! I think I will then hardcode the permalink in the `_data/technical_tasks.js` file.

Do you know if by any chance I can access Eleventy's functions link permalink computation? If this is possible, I could write something like:

```js
/ _data/recipes.js
const technical_tasks = require("./technical_tasks");
const { computePermalinkFor } = require('eleventy');

module.exports = [
  {
    title: 'Macarons',
    picture: 'macarons.jpg',
    steps: [
      {
        picture: 'step1.jpg',
        // NOTE: The backticks and JavaScript template string syntax here.
        content: `Take the eggs and [scramble](${ computePermalinkFor(technical_tasks.egg_scrambling) }) them.`
      },
    ]
  }
];
```

Or maybe at least do I have access to the slugify function?

Thank you!

**Reply by Bernardstanislas** [reply]
I think I found a flexible solution based on what you suggested. I can code myself a `computeTechnicalTaskPermalink` function, which I use in my `_data/technical_tasks.js` file and in my `_data/recipes.js`. This way I am sure the permalink to my technical task is the same in both files so my navigation will work without hardcoding.

**Reply by pdehaan** [reply]
I'm not familiar w/ `computePermalinkFor`, so you'd be in uncharted waters.

But you should be able to `require` Eleventy's built-in `slug` filter if you needed it (although not convinced you do, and that you want `slug` vs `url` here; you're already hardcoding URLs in ./_data/technical_tasks.js, why bother re-slugging or using other filters to tweak the URL).

```js
// _data/recipes.js
const slugFn = require("@11ty/eleventy/src/Filters/Slug");

const technical_tasks = require("./technical_tasks");

module.exports = [
  {
    title: 'Macarons',
    picture: 'macarons.jpg',
    steps: [
      {
        picture: 'step1.jpg',
        content: `Take the eggs and [scramble](${ slugFn(technical_tasks.egg_scrambling) }) them.`
      },
    ]
  }
];
```

But now by using `slugFn()`, it converts the hardcoded URL from `"/technical/scramble-eggs/"` to `[scramble](technicalscramble-eggs)` because now it thinks its a slug instead of a URL. So you might actually want the following: 

```js
const urlFn = require("@11ty/eleventy/src/Filters/Url");
```

I don't think that global data files have access to the `eleventyConfig` object like in the .eleventy.js config file, otherwise you could have used `eleventyConfig.getFilter("slug")`, if I recall correctly.

**Reply by pdehaan** [reply]
But I’d love to see your `computeTechnicalTaskPermalink` solution when you’ve got it all working.

**Reply by Bernardstanislas** [reply]
I think I landed something cool here: https://github.com/Bernardstanislas/salutchef.fr/tree/permalinks/src

Using your technique with a permalink computation function works very well, thank you!

**Reply by pdehaan** [reply]
Looks good! I think only thing that mildly concerned me was was that if you pass an `id` which isn't valid (typo or something changes), the `.find()` will probably fail and return `undefined` and then you're trying to access the `.permalink` on an `undefined` object.

```js
const getSkillPermalink = (id) => skills.find((skill) => skill.id === id).permalink;
```

It might not be an issue though, because if you mistype "macaron" it might break the 11ty build (trying to access `.permalink` on an `undefined` object), which might be a good workflow since it'd warn you early that there was an issue.

**Reply by Bernardstanislas** [reply]
Yes that's what I had in mind, if any typo occurs the build fails so I'll see it right away.

Moreover I plan on using typescript in the future to catch those typos.
